etk-asm/src/asm.rs

```diff
@@ -415,13 +570,16 @@ impl Assembler {
     }
 
     fn choose_sizes(&mut self) -> Result<(), Error> {
-        let minimum = Specifier::push_for(self.concrete_len).unwrap();
-
-        let mut undefined_labels: HashMap<String, Specifier> = self
-            .declared_labels
+        let mut sizes: HashMap<u64, Specifier> = self
+            .pending
             .iter()
-            .filter(|(_, v)| v.is_none())
-            .map(|(k, _)| (k.clone(), minimum))
+            .filter(|op| matches!(op, RawOp::Op(AbstractOp::Push(_))))
+            .map(|op| {
+                let expr = op.expr().unwrap();
+                let mut s = DefaultHasher::new();
+                expr.hash(&mut s);
+                (s.finish(), Specifier::push_for(1).unwrap())
```

Comment:
My assumption was that if there is a hash collision we can consider the expressions equivalent and therefore there should only be one entry for the expression in the hash map. Also, since they are equivalent, the specifier would need to be equivalent.

Would look something like this:
```rust
use std::collections::HashMap;

fn main() {
    let map: HashMap<_,_> = vec![(1,2), (2,3), (1,2), (3,4)].into_iter().collect();
    println!("{:?}", map);
}
```
```console
$ cargo run
{3: 4, 1: 2, 2: 3}
```

The only concern would be if an expression was hashed while it still has a variable `Terminal`, since the specifier for the expression would dependent on the value of the variable. I don't think that is an issue here because it's collecting over `self.pending` which always should always have either a pending macro invocation or have the pending instructions from an expanded macro (e.g. with the variables filled).

Does this answer your question or am I just rambling?

Reply:
Like I mentioned on telegram, my concern is more about the case where `hash(a) == hash(b)` but `a != b`.

Since you can put a 256-bit constant into an `Expression`, and a hash is only a `u64`, even with just plain numbers we have a huge number of collisions.

Reply:
Fixed